### DISCUSSION TITLE
Separate Directus tables from content

### DISCUSSION THREAD

**Discussion by Radiergummi** [directus/directus#2882] [Feature Requests]
As suggested by @benhaynes, I'd like to propose a new feature: Add the ability to create the Directus tables in a separate database to the actual content.

Lots of software applications use migrations to control their database schema. Migrations define all changes to the database schema in code and can be used to reproduce the database deterministically, as it has inevitably grown and changed over the lifetime of an application. As such, migrations must be the only means to modify the schema, else things go haywire quickly.  
These kinds of applications usually have a separate control panel application, so non-engineering folks can configure business aspects without having to resort to developers or SQL themselves: Say, updating pricing plans, or deleting old user accounts, or driving business reports.  
Creating this control panel usually binds lots of time and resources for a boring project that needs to continually be maintained and dragged along, and is loathed by both devs and users: Nobody pays attention to maintainability or UX, because "its just for internal use, don't waste too much time".

How awesome would it be if you could just drop in Directus instead to manage table _content_, with the schema being read-only?

So what I'm actually suggesting here is the possibility to add remote databases, creating the Directus tables in a designated Directus database. This way, Directus would be able to run self-contained on Docker, while being simply pointed to the content database hosting the data.  
This would allow to instantly start using Directus, even for huge or dynamic pre-existing data stores.

I'm well aware this might be too big of a step, or a direction you don't want to go into, but as an avid Directus user from a previous job, I thought I'd give it a spin 😊

**Comment by nvitaterna** [comment]
I'd love to see the same thing. It does make sense to separate configuration from content (IMO). Having a separate database purely for Directus would make migrations easier.

**Reply by rijkvanzanten** [reply]
The only tricky thing we have to solve for first is how to do relationships between the two different data sources. The system stuff might have to relate back for things like revisions. We should also consider moving users and files to be a user collection rather than system tables to really enforce that separation of concerns

**Reply by benhaynes** [reply]
I'm a big fan of moving the users/files to a "normal" collection instead of metadata.

**Reply by nvitaterna** [reply]
> The only tricky thing we have to solve for first is how to do relationships between the two different data sources. The system stuff might have to relate back for things like revisions.

What about a database field in directus_collections?

**Reply by rijkvanzanten** [reply]
> > The only tricky thing we have to solve for first is how to do relationships between the two different data sources. The system stuff might have to relate back for things like revisions.
> 
> 
> 
> What about a database field in directus_collections?

Yeah! Though that doesn't solve for doing the actual relational lookup for nested api responses etc. The real challenge here is to update the querying engine in such a way where it no longer relies on regular joins in a query, but can also dynamically query from other sources 👍🏻

**Reply by benhaynes** [reply]
I don't think defining those relationships is hard... I think he meant actually running performant SQL joins across different datastores. It will be hard to join virtualized relational data across different databases.

**Comment by henk23** [comment]
One approach could be kind of like Wordpress' ACF Pro solves it with their acf-json directory:

The database stays as it is, but additionally for each model/role/permission, a json file is saved somewhere in the filesystem, containing all the details to reproduce it. (Also giving dev teams the chance to check them into git.)

Whenever the schema changes, the json file changes. 

Each file has a "modified" timestamp and when directus discovers a newer json definition (for example because of a git pull), it will offer a sync button to incorporate the changes into the db schema.

Or if you prefer thinking in terms of migrations (because these json files can get messy): Each change to the schema creates a new migration file. (These can even be pure .sql files, so it would be possible to do fancy custom things in there.) Directus tracks which migrations have been run and when a new migration is discoverd, gives the developer an option to run it to update their database. (Or even run them automatically, analog to the EXTENSIONS_AUTO_RELOAD function, or on server start.)

**Reply by rijkvanzanten** [reply]
This is basically what we currently do through the schema snapshotting feature 🙂

**Reply by henk23** [reply]
Whoa, I've been working with directus for months now and never ever have I once looked into the "CLI" part of the docs. Thanks so much!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
